packages/shared/src/server/repositories/traces.ts

```diff
@@ -0,0 +1,159 @@
+import { TraceClickhouseRecord } from "../clickhouse/schema";
+import { queryClickhouse } from "./clickhouse";
+import {
+  createFilterFromFilterState,
+  getProjectIdDefaultFilter,
+} from "../queries/clickhouse-filter/factory";
+import { ObservationLevel } from "@prisma/client";
+import { FilterState } from "../../types";
+import { logger } from "../logger";
+
+export type TracesTableReturnType = Pick<
+  TraceClickhouseRecord,
+  | "project_id"
+  | "id"
+  | "name"
+  | "timestamp"
+  | "bookmarked"
+  | "release"
+  | "version"
+  | "user_id"
+  | "session_id"
+  | "tags"
+  | "metadata"
+  | "public"
+> & {
+  level: ObservationLevel;
+  observation_count: number | null;
+  latency: number;
+  usage_details: Record<string, number>;
+  cost_details: Record<string, number>;
+  scores_avg: Array<{ name: string; avg_value: number }>;
+};
+
+export const getTracesTable = async (
+  projectId: string,
+  filter: FilterState,
+  limit?: number,
+  offset?: number,
+) => {
+  logger.info(`input filter ${JSON.stringify(filter)}`);
+  const { tracesFilter, scoresFilter, observationsFilter } =
+    getProjectIdDefaultFilter(projectId, { tracesPrefix: "t" });
+
+  tracesFilter.push(
+    ...createFilterFromFilterState(filter, { tracesPrefix: "t" }),
+  );
+
+  const tracesFilterRes = tracesFilter.apply();
+  const scoresAvgFilterRes = scoresFilter.apply();
+  const observationsStatsRes = observationsFilter.apply();
+
+  const query = `
+  WITH observations_stats AS (
+  SELECT
+    COUNT(*) AS observation_count,
+      sumMap(usage_details) as usage_details,
+      SUM(total_cost) AS total_cost,
+      date_diff('seconds', least(min(start_time), min(end_time)), greatest(max(start_time), max(end_time))) as latencyMs,
+      multiIf(
+        arrayExists(x -> x = 'ERROR', groupArray(level)), 'ERROR',
+        arrayExists(x -> x = 'WARNING', groupArray(level)), 'WARNING',
+        arrayExists(x -> x = 'DEFAULT', groupArray(level)), 'DEFAULT',
+        'DEBUG'
+      ) AS level,
+      sumMap(cost_details) as cost_details,
+      trace_id,
+      project_id
+    FROM
+        observations final
```

Comment:
Those finals may become problematic - perhaps we can already have a backup plan ready in case we need to migrate off of them.

Reply:
We would need to group by and select the first one.